File: Cores/OSOL.Extremum.Core.DotNet/Optimization/Nodes/GeneralNode.cs
using System;
using System.Collections.Generic;

namespace OSOL.Extremum.Core.DotNet.Optimization.Nodes
{
    public abstract class GeneralNode<TBase, TFuncType, TV> where TV: class, IOptimizable<TBase, TFuncType>
    {
        public int NodeId;
        
        public abstract void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state);
        public abstract void Process(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state);

        public virtual int? GetCurrentCondition(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area,
            State<TBase, TFuncType, TV> state) => null;
    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Optimization/Nodes/GeneralNode.scala
package OSOL.Extremum.Cores.JVM.Optimization.Nodes

import OSOL.Extremum.Cores.JVM.Optimization._

abstract class GeneralNode[Base, FuncType, V <: Optimizable[Base, FuncType]](val nodeId: Int) {

  def initialize(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit

  def process(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit

  def getCurrentCondition(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Option[Int] = None

}
